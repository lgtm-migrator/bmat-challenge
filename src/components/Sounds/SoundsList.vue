<template>
  <div class="sounds-list">
    <table class="sounds-list__table">
      <thead>
        <tr>
          <th scope="col">ISWC</th>
          <th scope="col">Work Title</th>
          <th class="sounds-list__shares" scope="col">Shares</th>
          <th scope="col"></th>
          <th scope="col"></th>
        </tr>
      </thead>
      <tbody>
        <tr class="sounds-list__rows" v-for="sound in soundList" :key="sound.id">
          <td>{{ sound.iswc }}</td>
          <td>{{ sound.work_title }}</td>
          <td class="sounds-list__shares">{{ sound.share }}</td>
          <td class="sounds-list__icon-wrapper">
            <router-link
              class="sounds-list__link"
              :to="{
                name: 'SoundDetails',
                params: {
                  id: sound.id
                }
              }"
            >
              <button class="sounds-list__details-btn">
                &#9654;
              </button>
            </router-link>
          </td>
          <td class="sounds-list__icon-wrapper">
            <button @click="deleteSound(sound.id)" class="sounds-list__delete-btn">
              &#10006;
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapActions } from 'vuex'

export default {
  name: 'SoundsList',

  props: {
    soundList: Array
  },

  data () {
    return {
      sounds: []
    }
  },

  computed: {
    /**
     * @return {Array}
     */
    receivedSounds () {
      return this.soundList
    }
  },

  methods: {
    ...mapActions(['deleteById']),

    /**
     * Delete sound from the list
     * @param {number} id
     */
    deleteSound (id) {
      this.deleteById(id)
    }
  }
}
</script>

<style lang="scss" scoped>
.sounds-list {
  margin-top: $unit*4;
  padding: 0 $unit*4 $unit*6 $unit*4;
}

.sounds-list__table {
  border-bottom: 1px dashed $grey-light;
  border-top: 1px dashed $grey-light;
  border-collapse: collapse;
  box-shadow: 0 0 $unit*2 $grey-light;
  text-align: left;
  width: 100%;

  th,
  td {
    padding: $unit*4;
  }

  th {
    background: $brand-primary;
    color: $white;
  }
}

.sounds-list__info,
.sounds-list__icon-wrapper {
  text-align: center;
}

.sounds-list__icon-wrapper {
  width: 80px;
}

.sounds-list__rows {
  background: $selection-color;

  &:nth-of-type(odd) {
   background: $white;
  }
}

.sounds-list__shares {
  text-align: right;
}

.sounds-list__details-btn {
  background: $grey-lighter;
  border: none;
  border-radius: 4px;
  color: $white;
  font-style: unset;
  padding: $unit $unit*3;

  &:hover {
    background: $brand-primary;
    cursor: pointer;
    transition: background 0.5s ease-in-out;
  }
}

.sounds-list__delete-btn {
  border: none;
  border-radius: 4px;
  color: $danger;
  font-style: unset;
  padding: $unit $unit*3;

  &:hover {
    cursor: pointer;
  }
}
</style>
