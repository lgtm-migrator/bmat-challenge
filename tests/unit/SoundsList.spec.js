import Vuex from 'vuex'
import VueRouter from 'vue-router'
import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import SoundsList from '@/components/Sounds/SoundsList.vue'
import mockedSounds from '@/mocks/sound-recordings-mock.json'

const localVue = createLocalVue()
localVue.use(VueRouter)
localVue.use(Vuex)

let wrapper = null

const actions = {
  deleteById: jest.fn()
}

const mockDeleteById = actions.deleteById

beforeEach(() => {
  wrapper = shallowMount(SoundsList, {
    propsData: {
      soundList: mockedSounds
    },
    stubs: {
      RouterLink: RouterLinkStub
    },
    actions,
    localVue
  })
})

afterEach(() => {
  wrapper.destroy()
})

describe('SoundsList.vue', () => {
  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('has a table with a list of sounds', () => {
    const soundsListTable = wrapper.find('.sounds-list__table')
    expect(soundsListTable).toBeTruthy()
  })

  it('has 6 sounds on the list', () => {
    const soundsList = wrapper.findAll('.sounds-list__rows')
    expect(soundsList).toHaveLength(6)
  })

  it('triggers the delete sound button', () => {
    actions.deleteById(mockedSounds[3].id)

    expect(mockDeleteById).toHaveBeenCalled()
    expect(mockDeleteById).toHaveBeenCalledWith(mockedSounds[3].id)
  })

  it('triggers the sound details button', () => {
    expect(wrapper.findComponent(RouterLinkStub).props().to.params.id).toBe(mockedSounds[0].id)
  })
})
