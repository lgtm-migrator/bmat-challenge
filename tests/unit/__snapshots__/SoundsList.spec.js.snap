// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SoundsList.vue renders correctly 1`] = `
<div
  class="sounds-list"
>
  <table
    class="sounds-list__table"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          ISWC
        </th>
         
        <th
          scope="col"
        >
          Work Title
        </th>
         
        <th
          class="sounds-list__shares"
          scope="col"
        >
          Shares
        </th>
         
        <th
          scope="col"
        />
         
        <th
          scope="col"
        />
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="sounds-list__rows"
      >
        <td>
          T9204644531
        </td>
         
        <td>
          RAISE EM UP
        </td>
         
        <td
          class="sounds-list__shares"
        >
          50
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <a
            class="sounds-list__link"
          >
            <button
              class="sounds-list__details-btn"
            >
              
              ▶
            
            </button>
          </a>
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <button
            class="sounds-list__delete-btn"
          >
            
            ✖
          
          </button>
        </td>
      </tr>
      <tr
        class="sounds-list__rows"
      >
        <td>
          T9142925460
        </td>
         
        <td>
          A FIRE LOVE
        </td>
         
        <td
          class="sounds-list__shares"
        >
          88
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <a
            class="sounds-list__link"
          >
            <button
              class="sounds-list__details-btn"
            >
              
              ▶
            
            </button>
          </a>
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <button
            class="sounds-list__delete-btn"
          >
            
            ✖
          
          </button>
        </td>
      </tr>
      <tr
        class="sounds-list__rows"
      >
        <td>
          T0701815124
        </td>
         
        <td>
          TURTLE
        </td>
         
        <td
          class="sounds-list__shares"
        >
          100
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <a
            class="sounds-list__link"
          >
            <button
              class="sounds-list__details-btn"
            >
              
              ▶
            
            </button>
          </a>
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <button
            class="sounds-list__delete-btn"
          >
            
            ✖
          
          </button>
        </td>
      </tr>
      <tr
        class="sounds-list__rows"
      >
        <td>
          T0724517138
        </td>
         
        <td>
          THE LOVE
        </td>
         
        <td
          class="sounds-list__shares"
        >
          0
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <a
            class="sounds-list__link"
          >
            <button
              class="sounds-list__details-btn"
            >
              
              ▶
            
            </button>
          </a>
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <button
            class="sounds-list__delete-btn"
          >
            
            ✖
          
          </button>
        </td>
      </tr>
      <tr
        class="sounds-list__rows"
      >
        <td>
          T9032296961
        </td>
         
        <td>
          THE LOVE
        </td>
         
        <td
          class="sounds-list__shares"
        >
          0
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <a
            class="sounds-list__link"
          >
            <button
              class="sounds-list__details-btn"
            >
              
              ▶
            
            </button>
          </a>
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <button
            class="sounds-list__delete-btn"
          >
            
            ✖
          
          </button>
        </td>
      </tr>
      <tr
        class="sounds-list__rows"
      >
        <td>
          T9159537301
        </td>
         
        <td>
          THE LOVE
        </td>
         
        <td
          class="sounds-list__shares"
        >
          89.4565432344
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <a
            class="sounds-list__link"
          >
            <button
              class="sounds-list__details-btn"
            >
              
              ▶
            
            </button>
          </a>
        </td>
         
        <td
          class="sounds-list__icon-wrapper"
        >
          <button
            class="sounds-list__delete-btn"
          >
            
            ✖
          
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
